crates/polars-io/src/cloud/credential_provider.rs

```diff
@@ -0,0 +1,679 @@
+        struct S(
+            CredentialProviderFunction,
+            FetchedCredentialsCache<Arc<object_store::gcp::GcpCredential>>,
+        );
+
+        #[async_trait]
+        impl object_store::CredentialProvider for S {
+            type Credential = object_store::gcp::GcpCredential;
+
+            async fn get_credential(&self) -> object_store::Result<Arc<Self::Credential>> {
+                self.1
+                    .get_maybe_update(async {
+                        let (creds, expiry) = self.0 .0().await?;
+                        PolarsResult::Ok((creds.unwrap_gcp(), expiry))
+                    })
+                    .await
+                    .map_err(build_to_object_store_err!("credential-provider-gcp"))
+            }
+        }
+
+        Arc::new(S(
+            self,
+            FetchedCredentialsCache::new(Arc::new(GcpCredential {
+                bearer: String::new(),
+            })),
+        ))
+    }
+}
+
+impl Debug for CredentialProviderFunction {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "credential provider function at 0x{:016x}",
+            self.0.as_ref() as *const _ as *const () as usize
+        )
+    }
+}
+
+impl Eq for CredentialProviderFunction {}
+
+impl PartialEq for CredentialProviderFunction {
+    fn eq(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.0, &other.0)
+    }
+}
+
+impl Hash for CredentialProviderFunction {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        state.write_usize(Arc::as_ptr(&self.0) as *const () as usize)
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de> serde::Deserialize<'de> for PlCredentialProvider {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        #[cfg(feature = "python")]
+        {
+            Ok(Self::Python(PythonCredentialProvider::deserialize(
+                deserializer,
+            )?))
+        }
+        #[cfg(not(feature = "python"))]
+        {
+            use serde::de::Error;
+            Err(D::Error::custom("cannot deserialize PlCredentialProvider"))
+        }
+    }
+}
+
+#[cfg(feature = "serde")]
+impl serde::Serialize for PlCredentialProvider {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        use serde::ser::Error;
+
+        // TODO:
+        // * Add magic bytes here to indicate a python function
+        // * Check the Python version on deserialize
+        #[cfg(feature = "python")]
+        if let PlCredentialProvider::Python(v) = self {
+            return v.serialize(serializer);
+        }
+
+        Err(S::Error::custom(format!("cannot serialize {:?}", self)))
+    }
+}
+
+/// Avoids calling the credential provider function if we have not yet passed the expiry time.
+#[derive(Debug)]
+struct FetchedCredentialsCache<C>(tokio::sync::Mutex<(C, u64)>);
```

Comment:
Nice one!

Reply:
I tried changing it but couldn't as `object_store::CredentialProvider` requires `Send`, so we have to use tokio's mutex